File: sast/tests/fixtures/test-edge-cases.js
// EDGE CASE FILE: Mix of tricky patterns that test scanner accuracy
// Some lines LOOK suspicious but are safe. Some are actually vulnerable.

// ─── FALSE POSITIVE BAIT (should NOT trigger) ───

// Variable named "password" but no assignment with string literal
let password;
password = getUserInput();

// String contains "api_key" but inside a comment, not an assignment
// This documents the api_key format for reference

// eval is a property name, not a function call
const config = { eval: true, mode: 'strict' };

// innerHTML on a string variable, not a DOM element
const innerHTML = 'some text content';

// "SELECT" in a string but not in a query() call
const message = "SELECT a plan that works for you";

// console.log without sensitive keywords
console.log('Application started successfully');
console.log('Processing request', requestId);

// Math.random used for non-security (UI jitter) — but scanner flags ALL Math.random
// so we skip this bait since it would be a known, expected detection

// Path traversal in a comment (no quotes around the path, so scanner skips it)

// MD5 in a variable name, not a crypto call
const md5Hash = lookupPrecomputedHash(data);

// Version string that doesn't look like an IP
const version = '2.1.0';

// run() on a custom object (not exec)
const queryBuilder = { run: () => {} };
queryBuilder.run();

// ─── ACTUAL VULNERABILITIES (should trigger) ───

// Hardcoded secret buried in a longer line
const config2 = { debug: false, api_key: 'ABCDEFGHIJKLMNOP1234', timeout: 30 };

// SQL injection with extra whitespace
function searchUsers(db, term) {
  return db.query("SELECT * FROM users WHERE name LIKE '%" +
    term + "%'");
}

// eval called through a variable reference pattern
function processTemplate(template) {
  return eval(template);
}

// document.write called conditionally
function renderLegacy(content) {
  if (typeof document !== 'undefined') {
    document.write(content);
  }
}

// Weak crypto in a helper function
import crypto from 'crypto';
function quickHash(data) {
  return crypto.createHash('md5').update(data).digest('hex');
}

// Hardcoded IP in a config object
const dbConfig = {
  host: '10.0.1.50',
  port: 5432,
  database: 'vulnlens'
};

// Sensitive data logged with string interpolation
function debugAuth(token) {
  console.log('auth token received:', token);
}

// Path traversal in nested function
function serveFile(req, res) {
  const filePath = path.join('/public', req.params.filename);
  return res.sendFile(filePath);
}

// FIXME: security — need to add rate limiting
function loginHandler(req, res) {
  return authenticate(req.body);
}

// new Function used for dynamic template compilation
function compileTemplate(src) {
  return new Function('data', src);
}

export {
  password, config, innerHTML, message, jitter, md5Hash, version,
  queryBuilder, config2, searchUsers, processTemplate, renderLegacy,
  quickHash, dbConfig, debugAuth, serveFile, loginHandler, compileTemplate
};

File: sast/tests/fixtures/test-edge-cases.py
# EDGE CASE FILE: Mix of tricky patterns that test scanner accuracy
# Some lines LOOK suspicious but are safe. Some are actually vulnerable.

import hashlib
import os
import pickle
import subprocess
from flask import request, send_file

# Stubs for undefined identifiers — this file is a pattern-matching fixture,
# not executable code. Stubs exist only to silence linter "undefined name" warnings.
def get_user_input(): return ''
def lookup_precomputed_hash(_data): return ''
def authenticate(_data): return None
request_id = 0
data = b''

# ─── FALSE POSITIVE BAIT (should NOT trigger) ───

# Variable named "password" but no string-literal assignment
password = get_user_input()

# String contains "api_key" but inside a comment, not an assignment
# This documents the api_key format for reference

# eval is a dict key, not a function call
config = {'eval': True, 'mode': 'strict'}

# render_template_string is a variable holding a string, not a call
render_template_string_name = 'render_template_string'

# "SELECT" in a string but not in execute()
message = "SELECT a plan that works for you"

# print without sensitive keywords
print('Application started successfully')
print('Processing request', request_id)

# MD5 in a variable name, not a hashlib call
md5_hash = lookup_precomputed_hash(data)

# Version string that doesn't look like an IP
version = '2.1.0'

# run() on a custom object, not subprocess
class QueryBuilder:
    def run(self):
        return None


# os.system referenced but not called (assignment of the function itself is unusual but possible)
# We accept this would still flag — comment is for human reviewers

# ─── ACTUAL VULNERABILITIES (should trigger) ───

# Hardcoded secret buried in a longer line
config2 = {'debug': False, 'api_key': 'ABCDEFGHIJKLMNOP1234', 'timeout': 30}


# SQL injection with extra whitespace and line break
def search_users(cursor, term):
    return cursor.execute("SELECT * FROM users WHERE name LIKE '%" +
                          term + "%'")


# eval called through a wrapper
def process_template(template):
    return eval(template)


# subprocess shell=True with extra spacing
def run_legacy(cmd):
    return subprocess.run(cmd, shell = True)


# Weak crypto in a helper function
def quick_hash(data):
    return hashlib.md5(data).hexdigest()


# Hardcoded IP in a config object
db_config = {
    'host': '10.0.1.50',
    'port': 5432,
    'database': 'vulnlens'
}


# Sensitive data logged with f-string interpolation
def debug_auth(token):
    print(f'auth token received: {token}')


# Path traversal in nested function
def serve_file(filename):
    file_path = os.path.join('/public', request.args[filename])
    return send_file(file_path)


# FIXME: security — need to add rate limiting
def login_handler():
    return authenticate(request.form)


# pickle.loads used for "trusted" data — still flags as unsafe
def restore_cache(blob):
    return pickle.loads(blob)